flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/stopwithsavepoint/StopWithSavepointOperationManager.java

```diff
@@ -45,18 +57,28 @@ public StopWithSavepointTerminationManager(
      * Enforces the correct completion order of the passed {@code CompletableFuture} instances in
      * accordance to the contract of {@link StopWithSavepointTerminationHandler}.
      *
-     * @param completedSavepointFuture The {@code CompletableFuture} of the savepoint creation step.
+     * @param terminate Flag indicating whether to terminate or suspend the job.
+     * @param targetDirectory Target for the savepoint.
      * @param terminatedExecutionStatesFuture The {@code CompletableFuture} of the termination step.
      * @param mainThreadExecutor The executor the {@code StopWithSavepointTerminationHandler}
      *     operations run on.
      * @return A {@code CompletableFuture} containing the path to the created savepoint.
      */
-    public CompletableFuture<String> stopWithSavepoint(
-            CompletableFuture<CompletedCheckpoint> completedSavepointFuture,
+    public CompletableFuture<String> trackStopWithSavepoint(
+            boolean terminate,
```

Comment:
Does it make sense to introduce an enum to give this flag a bit more expressive power? For example only looking at the code if one passes `false` then it is not very clear to me what this means.

Reply:
Yes! I believe this is necessary. The argument order is different in various parts of the codebase, and the semantics of the boolean flag are defined different between flink-runtime and flink-core.
However, I would like to tackle this in a follow up PR, as this change alone will be difficult to review (I started doing it, but realized that this is touching a lot of classes): https://issues.apache.org/jira/browse/FLINK-21638

```
 flink-clients/src/main/java/org/apache/flink/client/deployment/application/EmbeddedJobClient.java                           |  5 +++--
 flink-core/src/main/java/org/apache/flink/core/execution/JobClient.java                                                     |  6 +++---
 flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java                                  |  6 +++---
 flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointProperties.java                                   | 17 ++++++++---------
 flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/StopWithSavepointOperations.java                            | 12 +++++++++++-
 flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java                                             |  6 ++++--
 flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobMaster.java                                               |  7 +++++--
 flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobMasterGateway.java                                        |  3 ++-
 flink-runtime/src/main/java/org/apache/flink/runtime/minicluster/MiniCluster.java                                           |  5 +++--
 flink-runtime/src/main/java/org/apache/flink/runtime/minicluster/MiniClusterJobClient.java                                  |  5 +++--
 flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/savepoints/SavepointHandlers.java                     |  9 +++++++--
 flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerBase.java                                           |  9 +++++----
 flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerNG.java                                             |  4 +++-
 flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/AdaptiveScheduler.java                              |  9 +++++----
 flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/Executing.java                                      |  7 ++++---
 flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/StopWithSavepoint.java                              | 17 +++++++++--------
 flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/stopwithsavepoint/StopWithSavepointOperationManager.java     |  8 +++++---
 flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/RestfulGateway.java                                         |  5 +++--
 flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorTest.java                              |  3 ++-
 flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/TestingStopWithSavepointOperations.java                     |  2 +-
 flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/ExecutingTest.java                                  |  4 +++-
 flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/StopWithSavepointTest.java                          |  7 ++++---
 flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/stopwithsavepoint/StopWithSavepointOperationManagerTest.java |  5 +++--
 23 files changed, 99 insertions(+), 62 deletions(-)
 ```

Reply:
Sounds good.